File: Problems/437_Path_Sum_III.cpp
#include <unordered_map>
#include<vector>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// pass by reference makes it faster
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        int cnt = 0;
        // {pastSum, count}
        unordered_map<long long, int> seen;
        seen[0] = 1;
        traverse(root, targetSum, seen, cnt, 0);
        return cnt;
    }
    void traverse(TreeNode* node, int targetSum, unordered_map<long long, int>& seen, int& cnt, long long currSum) {
        if (!node) return;
        currSum += node -> val;
        cnt += seen[currSum - targetSum];
        ++seen[currSum];
        traverse(node -> left, targetSum, seen, cnt, currSum);
        traverse(node -> right, targetSum, seen, cnt, currSum);
        --seen[currSum];
    }
};

// Only faster than 5% of submissions
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        int cnt = 0;
        // {pastSum, count}
        unordered_map<long long, int> seen;
        seen[0] = 1;
        traverse(root, targetSum, seen, cnt, 0);
        return cnt;
    }
    void traverse(TreeNode* node, int targetSum, unordered_map<long long, int> seen, int& cnt, long long currSum) {
        if (!node) return;
        currSum += node -> val;
        cnt += seen[currSum - targetSum];
        ++seen[currSum];
        traverse(node -> left, targetSum, seen, cnt, currSum);
        traverse(node -> right, targetSum, seen, cnt, currSum);
    }
};

class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        int cnt = 0;
        vector<long> tmp = {0};
        traverse(root, targetSum, tmp, cnt);
        return cnt;
    }
    void traverse(TreeNode* node, int targetSum, vector<long> seen, int& cnt) {
        if (!node) return;
        long currSum = seen[seen.size()-1] + node -> val;
        for (long num : seen) {
            if (currSum - num == targetSum)
                ++cnt;
        }
        seen.push_back(currSum);
        traverse(node -> left, targetSum, seen, cnt);
        traverse(node -> right, targetSum, seen, cnt);
    }
};


// TLE
// class Solution {
// public:
//     int pathSum(TreeNode* root, int targetSum) {
//         int cnt = 0;
//         // presum, count
//         unordered_map<long long, int> tmp;
//         tmp[0] = 1;
//         traverse(root, cnt, 0, tmp, targetSum);
//         return cnt;
//     }
//     void traverse(TreeNode* node, int &cnt, long long currSum, unordered_map<long long, int>preSums, int targetSum){
//         if(!node) return;
//         currSum += node->val;
//         cnt += preSums[currSum-targetSum];
//         preSums[currSum]++;

//         traverse(node->left,  cnt, currSum, preSums, targetSum);
//         traverse(node->right, cnt, currSum, preSums, targetSum);
//     }
// };

File: Problems/437_Path_Sum_III.py
from collections import defaultdict
from typing import Optional, List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# use dict makes it faster
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        cnt = 0
        hm = defaultdict(int)
        hm[0] += 1
        def traverse(node, currSum):
            nonlocal cnt
            if not node:
                return
            currSum += node.val
            cnt += hm[currSum - targetSum]
            hm[currSum] += 1
            traverse(node.left, currSum)
            traverse(node.right, currSum)
            hm[currSum] -= 1
        traverse(root , 0)
        return cnt

# use nonlocal
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        cnt = 0
        def traverse(node, seen : List[int]):
            nonlocal cnt
            if not node:
                return
            for num in seen:
                if node.val + seen[-1] - num == targetSum:
                    cnt += 1
            traverse(node.left, seen + [node.val + seen[-1]])
            traverse(node.right, seen + [node.val + seen[-1]])               
        traverse(root , [0])
        return cnt

# use self
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        self.cnt = 0
        def traverse(node, seen : List[int]):
            if not node:
                return
            for num in seen:
                if node.val + seen[-1] - num == targetSum:
                    self.cnt += 1
            traverse(node.left, seen + [node.val + seen[-1]])
            traverse(node.right, seen + [node.val + seen[-1]])               
        traverse(root , [0])
        return self.cnt